rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -107,289 +113,144 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
-                                        *e,
-                                    ),
-                                );
-                            }
-                        }
+                .then_some(log.offset_id)
+            }));
+
+        // Merge the offset ids, assuming the user_offset_ids and filtered_offset_ids are ordered.
+        // This should be a sorted list of unique offset ids.
+        // We are treating the sorted vecs as sets.
+        let merged_oids_holder: Vec<u32>;
+        let merged_offset_ids = match (&input.filtered_offset_ids, &input.user_offset_ids) {
+            (Some(fids), Some(uids)) => {
+                // We take the set intersect of offset ids of those passing the filter and those correspond to user specified ids.
+                merged_oids_holder = merge_sorted_vecs_conjunction(fids, uids);
+                &merged_oids_holder
+            }
+            (Some(oids), None) | (None, Some(oids)) => oids,
+            _ => {
+                // If both filter and user offset ids are None, it suggests user did not specify any filter. We should return all available documents.
+                // Offset ids from the materialized records should be unique, so we only need to sort them.
+                let mut live_log_offset_ids = mat_records
+                    .iter()
+                    .map(|(log, _)| log.offset_id)
+                    .collect::<Vec<_>>();
+                live_log_offset_ids.sort();
+                merged_oids_holder = match &record_segment_reader {
+                    Some(reader) => {
+                        // TODO: We may not need to have the entire offset ids, given limit and offset.
+                        let compact_offset_ids =
+                            reader.get_all_offset_ids().await.map_err(|e| {
+                                tracing::error!("Error reading record segment: {}", e);
+                                MergeMetadataResultsOperatorError::RecordSegmentReadError
+                            })?;
+                        // We take the set union of offset ids from the log and from the record segment.
+                        merge_sorted_vecs_disjunction(&live_log_offset_ids, &compact_offset_ids)
                     }
+                    None => live_log_offset_ids,
                 };
-                for merged_id in merged_ids {
-                    // Skip already taken from log.
-                    if visited_ids.contains(&merged_id) {
-                        continue;
-                    }
+                &merged_oids_holder
+            }
+        }
+        .iter()
+        .filter(|oid| !deleted_offset_ids.contains(oid))
+        // Truncate the offset ids using offset and limit.
+        // The .skip(...) and .take(...) will return an empty iterator if the offset and limit is out of bound.
+        .skip(input.offset.unwrap_or(u32::MIN) as usize)
```

Comment:
For my own knowledge

skip relies on nth - https://github.com/rust-lang/rust/blob/1.45.0/src/libcore/iter/adapters/mod.rs#L2007

Which is done by pointer arithmetic for vecs

- https://github.com/rust-lang/rust/blob/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libcore/slice/mod.rs#L3253

so this is efficient